# S02 T05: Exploració de les dades


## Nivel 1

### Ejercicio 1
Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un pandas Dataframe. Explora les dades que conté, i queda’t únicament amb les columnes que consideris rellevants.



**Análisis general de datos del Data set**

In [1]:
import pandas as pd
import numpy as np


delayFlights = pd.read_csv('DelayedFlights.csv', index_col = 0)

delayFlights

/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0        2008      1           3          4   2003.0        1955   2211.0   
1        2008      1           3          4    754.0         735   1002.0   
2        2008      1           3          4    628.0         620    804.0   
4        2008      1           3          4   1829.0        1755   1959.0   
5        2008      1           3          4   1940.0        1915   2121.0   
...       ...    ...         ...        ...      ...         ...      ...   
7009710  2008     12          13          6   1250.0        1220   1617.0   
7009717  2008     12          13          6    657.0         600    904.0   
7009718  2008     12          13          6   1007.0         847   1149.0   
7009726  2008     12          13          6   1251.0        1240   1446.0   
7009727  2008     12          13          6   1110.0        1103   1413.0   

         CRSArrTime UniqueCarrier  FlightNum  ... TaxiIn  TaxiOut  Cancelled  \
0              2225            WN        335  ...    4.0      8.0          0   
1              1000            WN       3231  ...    5.0     10.0          0   
2               750            WN        448  ...    3.0     17.0          0   
4              1925            WN       3920  ...    3.0     10.0          0   
5              2110            WN        378  ...    4.0     10.0          0   
...             ...           ...        ...  ...    ...      ...        ...   
7009710        1552            DL       1621  ...    9.0     18.0          0   
7009717         749            DL       1631  ...   15.0     34.0          0   
7009718        1010            DL       1631  ...    8.0     32.0          0   
7009726        1437            DL       1639  ...   13.0     13.0          0   
7009727        1418            DL       1641  ...    8.0     11.0          0   

         CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                       N         0           NaN          NaN      NaN   
1                       N         0           NaN          NaN      NaN   
2                       N         0           NaN          NaN      NaN   
4                       N         0           2.0          0.0      0.0   
5                       N         0           NaN          NaN      NaN   
...                   ...       ...           ...          ...      ...   
7009710                 N         0           3.0          0.0      0.0   
7009717                 N         0           0.0         57.0     18.0   
7009718                 N         0           1.0          0.0     19.0   
7009726                 N         0           NaN          NaN      NaN   
7009727                 N         0           NaN          NaN      NaN   

         SecurityDelay  LateAircraftDelay  
0                  NaN                NaN  
1                  NaN                NaN  
2                  NaN                NaN  
4                  0.0               32.0  
5                  NaN                NaN  
...                ...                ...  
7009710            0.0               22.0  
7009717            0.0                0.0  
7009718            0.0               79.0  
7009726            NaN                NaN  
7009727            NaN                NaN  

[1936758 rows x 29 columns]

**Ver tipos de datos:**

In [2]:
delayFlights.dtypes

Year                   int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
DepTime              float64
CRSDepTime             int64
ArrTime              float64
CRSArrTime             int64
UniqueCarrier         object
FlightNum              int64
TailNum               object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin                object
Dest                  object
Distance               int64
TaxiIn               float64
TaxiOut              float64
Cancelled              int64
CancellationCode      object
Diverted               int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtype: object

**Ver el resumen por defecto de la función .describe**

In [44]:
#quita la notación científica
pd.options.display.float_format = '{:,.1f}'.format 


In [4]:
delayFlights.describe()

Year       Month  DayofMonth   DayOfWeek     DepTime  CRSDepTime  \
count 1,936,758.0 1,936,758.0 1,936,758.0 1,936,758.0 1,936,758.0 1,936,758.0   
mean      2,008.0         6.1        15.8         4.0     1,518.5     1,467.5   
std           0.0         3.5         8.8         2.0       450.5       424.8   
min       2,008.0         1.0         1.0         1.0         1.0         0.0   
25%       2,008.0         3.0         8.0         2.0     1,203.0     1,135.0   
50%       2,008.0         6.0        16.0         4.0     1,545.0     1,510.0   
75%       2,008.0         9.0        23.0         6.0     1,900.0     1,815.0   
max       2,008.0        12.0        31.0         7.0     2,400.0     2,359.0   

          ArrTime  CRSArrTime   FlightNum  ActualElapsedTime  ...    Distance  \
count 1,929,648.0 1,936,758.0 1,936,758.0        1,928,371.0  ... 1,936,758.0   
mean      1,610.1     1,634.2     2,184.3              133.3  ...       765.7   
std         548.2       464.6     1,944.7               72.1  ...       574.5   
min           1.0         0.0         1.0               14.0  ...        11.0   
25%       1,316.0     1,325.0       610.0               80.0  ...       338.0   
50%       1,715.0     1,705.0     1,543.0              116.0  ...       606.0   
75%       2,030.0     2,014.0     3,422.0              165.0  ...       998.0   
max       2,400.0     2,400.0     9,742.0            1,114.0  ...     4,962.0   

           TaxiIn     TaxiOut   Cancelled    Diverted  CarrierDelay  \
count 1,929,648.0 1,936,303.0 1,936,758.0 1,936,758.0   1,247,488.0   
mean          6.8        18.2         0.0         0.0          19.2   
std           5.3        14.3         0.0         0.1          43.5   
min           0.0         0.0         0.0         0.0           0.0   
25%           4.0        10.0         0.0         0.0           0.0   
50%           6.0        14.0         0.0         0.0           2.0   
75%           8.0        21.0         0.0         0.0          21.0   
max         240.0       422.0         1.0         1.0       2,436.0   

       WeatherDelay    NASDelay  SecurityDelay  LateAircraftDelay  
count   1,247,488.0 1,247,488.0    1,247,488.0        1,247,488.0  
mean            3.7        15.0            0.1               25.3  
std            21.5        33.8            2.0               42.1  
min             0.0         0.0            0.0                0.0  
25%             0.0         0.0            0.0                0.0  
50%             0.0         2.0            0.0                8.0  
75%             0.0        15.0            0.0               33.0  
max         1,352.0     1,357.0          392.0            1,316.0  

[8 rows x 24 columns]

**Ver lo valores únicos de cada serie**

In [5]:
delayFlights.nunique()

Year                    1
Month                  12
DayofMonth             31
DayOfWeek               7
DepTime              1438
CRSDepTime           1207
ArrTime              1440
CRSArrTime           1364
UniqueCarrier          20
FlightNum            7499
TailNum              5366
ActualElapsedTime     673
CRSElapsedTime        515
AirTime               650
ArrDelay             1128
DepDelay             1058
Origin                303
Dest                  304
Distance             1419
TaxiIn                180
TaxiOut               332
Cancelled               2
CancellationCode        4
Diverted                2
CarrierDelay          983
WeatherDelay          599
NASDelay              574
SecurityDelay         156
LateAircraftDelay     564
dtype: int64

**Las variables Cancelled y Diverted pueden ser transformadas a categóricas porque solo tienen dos tipos de datos. Una ventaja de trabajar con variables categoricas es que reducimos el tamaño de uso de la memoria RAM y el tamaño del archivo en si.**



In [6]:
delayFlights[['Cancelled', 'Diverted']] = delayFlights[['Cancelled', 'Diverted']].astype('category')

In [7]:
#comprobamos que se transformó
delayFlights.dtypes

Year                    int64
Month                   int64
DayofMonth              int64
DayOfWeek               int64
DepTime               float64
CRSDepTime              int64
ArrTime               float64
CRSArrTime              int64
UniqueCarrier          object
FlightNum               int64
TailNum                object
ActualElapsedTime     float64
CRSElapsedTime        float64
AirTime               float64
ArrDelay              float64
DepDelay              float64
Origin                 object
Dest                   object
Distance                int64
TaxiIn                float64
TaxiOut               float64
Cancelled            category
CancellationCode       object
Diverted             category
CarrierDelay          float64
WeatherDelay          float64
NASDelay              float64
SecurityDelay         float64
LateAircraftDelay     float64
dtype: object

**Descripción del significado de cada columna (extraído de otro usuario de Kaggle):**


1. Year 2008
2. Month 1-12
3. DayofMonth 1-31
4. DayOfWeek 1 (Monday) - 7 (Sunday)
5. DepTime actual departure time (local, hhmm)
6. CRSDepTime scheduled departure time (local, hhmm)
7. ArrTime actual arrival time (local, hhmm)
8. CRSArrTime scheduled arrival time (local, hhmm)
9. UniqueCarrier unique carrier code
10. FlightNum flight number
11. TailNum plane tail number: aircraft registration, unique aircraft identifier
12. ActualElapsedTime in minutes
13. CRSElapsedTime in minutes
14. AirTime in minutes
15. ArrDelay arrival delay, in minutes: A flight is counted as "on time" if it operated less than 15 minutes later the scheduled time shown in the carriers' Computerized Reservations Systems (CRS).
16. DepDelay departure delay, in minutes
17. Origin origin IATA airport code
18. Dest destination IATA airport code
19. Distance in miles
20. TaxiIn taxi in time, in minutes
21. TaxiOut taxi out time in minutes
22. Cancelled *was the flight cancelled
23. CancellationCode reason for cancellation (A = carrier, B = weather, C = NAS, D = security)
24. iverted 1 = yes, 0 = no
25. CarrierDelay in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.
26. WeatherDelay in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.
27. NASDelay in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.
28. SecurityDelay in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.
29. LateAircraftDelay in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagatio

[Fuente](https://www.kaggle.com/code/adveros/flight-delay-eda-exploratory-data-analysis?scriptVersionId=1821484&cellId=2)

• Année 2008
• Mois 1-12
• JourduMois 1-31
• DayOfWeek 1 (lundi) - 7 (dimanché)
• Heure de départ réelle DepTime (locale, hhmm)
• Heure de départ prévue CRSDepTime (locale, hhmm)
• ArrTime heure d'arrivée réelle (locale, hhmm)
• CRSArrTime heure d'arrivée définie (locale, hhmm)
• Code de transporteur unique UniqueCarrier
• Numéro de vol FlightNum
• Numéro de file d'attente de l'avion TailNum : immatriculation de l'avion, identifiant unique de l'avion
• ActualElapsedTime en minutes
• CRSElapsedTime en minutes
• Temps d'antenne en minutes
• Retard d'arrivée ArrDelay, en minutes : Un vol est compté comme « à l'heure » s'il a fonctionné moins de 15 minutes après l'heure indiquée dans les systèmes de réservation informatisés (CRS) des transporteurs.
• Délai de départ DepDelay, en minutes
• Code d'aéroport IATA d'origine
• Code de destination IATA de l'aéroport
• Distance en miles
• TaxiEn taxi dans le temps, en minutes
• TaxiOut temps de taxi en minutes
• Annulé *le vol a-t-il été annulé ?
• Code d'annulation raison de l'annulation (A = transporteur, B = météo, C = NAS, D = sécurité)
• Renvoi 1 = oui, 0 = non
• CarrierDelay en minutes : le retard du transporteur est sous le contrôle du transporteur aérien. Exemples d'événements pouvant déterminer le retard du transporteur : nettoyage de l'aéronef, dommages à l'aéronef, attente de l'arrivée des passagers ou de l'équipage en correspondance, bagages, impact d'oiseau, chargement de la cargaison, restauration, ordinateur, équipement du transporteur en panne, légalité de l'équipage (repos du pilote ou de l'accompagnateur) , dommages causés par des marchandises dangereuses, inspection technique, ravitaillement en carburant, prise en charge des passagers handicapés, équipage en retard , entretien des toilettes, maintenance, surventes, service d'eau potable, évacuation d'un passager indiscipliné, embarquement ou siège prêté, rangement des bagages à main, retards de poids et centrage.
• WeatherDelay en minutes : le délai météo est augmenté par des conditions météorologiques extrêmes ou dangereuses qui sont supposées ou se manifester au point de départ, en route ou au point d'arrivée.
• NASRetard en minutes : un retard sous le contrôle du système d'espace aérien national (NAS) peut inclure : des conditions météorologiques non extrêmes, des opérations aéroportuaires, un volume de trafic élevé, le contrôle du trafic aérien, etc.
• Délai de sécurité en minutes : le délai de sécurité est renforcé par l'évacuation d'un terminal ou d'un hall, le réembarquement d'un avion en raison d'une brèche de sécurité, un équipement de contrôle inopérant et/ou de longues files d'attente de plus de 29 minutes dans les zones de contrôle.
• LateAircraftDelay en minutes : retard d'arrivée à un aéroport dû à l'arrivée tardive du même avion à un aéroport précédent. L'effet d'entraînement d'un retard antérieur aux aéroports en aval est appelé propagation du retard.
>

**Elimino columnas que no son necesaria para el presente análisis**
Year: porque todo el Data Ser es para el mismo (2008)
También elimino día del mes y día de la semana porque no haré análisis sobre los mismos, no viendo que sean útiles de momento.
FlightNum: no utilizaré el dato del nro. de vuelo


In [8]:
delayFlights.drop(['Year', 'DayofMonth', 'DayOfWeek','FlightNum', 'TailNum',], axis=1, inplace=True)

delayFlights.head()

Month  DepTime  CRSDepTime  ArrTime  CRSArrTime UniqueCarrier  \
0      1  2,003.0        1955  2,211.0        2225            WN   
1      1    754.0         735  1,002.0        1000            WN   
2      1    628.0         620    804.0         750            WN   
4      1  1,829.0        1755  1,959.0        1925            WN   
5      1  1,940.0        1915  2,121.0        2110            WN   

   ActualElapsedTime  CRSElapsedTime  AirTime  ArrDelay  ...  TaxiIn TaxiOut  \
0              128.0           150.0    116.0     -14.0  ...     4.0     8.0   
1              128.0           145.0    113.0       2.0  ...     5.0    10.0   
2               96.0            90.0     76.0      14.0  ...     3.0    17.0   
4               90.0            90.0     77.0      34.0  ...     3.0    10.0   
5              101.0           115.0     87.0      11.0  ...     4.0    10.0   

  Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0         0                 N         0           NaN          NaN      NaN   
1         0                 N         0           NaN          NaN      NaN   
2         0                 N         0           NaN          NaN      NaN   
4         0                 N         0           2.0          0.0      0.0   
5         0                 N         0           NaN          NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
4           0.0               32.0  
5           NaN                NaN  

[5 rows x 24 columns]

### Ejercicio 2
Fes un informe complet del data set:.

- Resumeix estadísticament les columnes d’interès
-  Troba quantes dades faltants hi ha per columna
-  Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...)
-  Taula de les aerolínies amb més endarreriments acumulats
-  Quins són els vols més llargs? I els més endarrerits?

#### Analisis de datos de las comunas de interés

**Análisis de vuelos que no arribaron a tiempo (cancelaciones, suspensiones y diferentes tipos de retrasos)**

**Ver número y porcentaje de vuelos cancelados y desviados**

In [9]:
#ver la cantidad de vuelos cancelados
delayFlights['Cancelled'].value_counts()

def porcentaje(problema, totalVuelos):
    porcentaje = (problema/totalVuelos) * 100
    return (porcentaje)

print("El porcentaje de vuelos cancelados es ",porcentaje(633,1929004))



El porcentaje de vuelos cancelados es  0.03281486197021883


**Vuelos cancelados: 633 de 1936758
Es un 0,033 de los vuelos**
    

In [10]:
#ver la cantidad de vuelos desviado
delayFlights['Diverted'].value_counts()

print("El porcentaje de vuelos deviados es ",porcentaje(7754,1929004))


    

El porcentaje de vuelos deviados es  0.4019690990791102


**Vuelos desviados: 7754 de 1936758
Es un 0,040 de los vuelos**


**Cantidad de minutos perdidos por cada una de las cuatro causas de retraso y suma de ellas**

In [11]:
#CarrierDelay, WeatherDelay, NASDelay y LateAircraftDelay.
analisysDeley = ['CarrierDelay', 'WeatherDelay', 'NASDelay', 'LateAircraftDelay']
delayFlights[analisysDeley].sum()
                

CarrierDelay        23,926,070.0
WeatherDelay         4,620,160.0
NASDelay            18,739,310.0
LateAircraftDelay   31,557,038.0
dtype: float64

Si bien existe un indicador de minutos de retraso (ArrDelay) este suma la totalidad de minutos de retraso cuando en realidad se considera que un vuelo tuvo una real demora a partir de los 15 minutos. Por eso resulta útil sumar todas los minutos que se perdieron en vuelos considerados retrasados.
Otra forma de hacerlo hubiera sido sumar todos los valores de la calomuna ArrDelay mayores o iguales a 15.

In [12]:
print("La total de minutos perdidos en vuelos demorados por las diferentes causas es de ",
      delayFlights['CarrierDelay'].sum() + delayFlights['WeatherDelay'].sum() + 
      delayFlights['NASDelay'].sum() + 
      delayFlights['LateAircraftDelay'].sum())

La total de minutos perdidos en vuelos demorados por las diferentes causas es de  78842578.0


**Media de minutos perdidos por cada causa**

In [13]:
delayFlights[analisysDeley].mean()


CarrierDelay        19.2
WeatherDelay         3.7
NASDelay            15.0
LateAircraftDelay   25.3
dtype: float64

In [14]:

delayFlights[analisysDeley].describe()


CarrierDelay  WeatherDelay    NASDelay  LateAircraftDelay
count   1,247,488.0   1,247,488.0 1,247,488.0        1,247,488.0
mean           19.2           3.7        15.0               25.3
std            43.5          21.5        33.8               42.1
min             0.0           0.0         0.0                0.0
25%             0.0           0.0         0.0                0.0
50%             2.0           0.0         2.0                8.0
75%            21.0           0.0        15.0               33.0
max         2,436.0       1,352.0     1,357.0            1,316.0

**Voy a crear un nuevo DataFrame solo con los datos de vuelos demorados para manipularlos**

In [15]:
onlyDelay = pd.DataFrame().assign(CarrierDelay = delayFlights['CarrierDelay'], 
                                  WeatherDelay = delayFlights['WeatherDelay'], 
                                  NASDelay = delayFlights['NASDelay'], 
                                  LateAircraftDelay = delayFlights['LateAircraftDelay'],
                                  ArrDelay = delayFlights['ArrDelay'])

onlyDelay

CarrierDelay  WeatherDelay  NASDelay  LateAircraftDelay  ArrDelay
0                 NaN           NaN       NaN                NaN     -14.0
1                 NaN           NaN       NaN                NaN       2.0
2                 NaN           NaN       NaN                NaN      14.0
4                 2.0           0.0       0.0               32.0      34.0
5                 NaN           NaN       NaN                NaN      11.0
...               ...           ...       ...                ...       ...
7009710           3.0           0.0       0.0               22.0      25.0
7009717           0.0          57.0      18.0                0.0      75.0
7009718           1.0           0.0      19.0               79.0      99.0
7009726           NaN           NaN       NaN                NaN       9.0
7009727           NaN           NaN       NaN                NaN      -5.0

[1936758 rows x 5 columns]

In [16]:
#Borro las filas con datos vacios
onlyDelay.dropna(inplace = True)

onlyDelay.describe()

CarrierDelay  WeatherDelay    NASDelay  LateAircraftDelay    ArrDelay
count   1,247,488.0   1,247,488.0 1,247,488.0        1,247,488.0 1,247,488.0
mean           19.2           3.7        15.0               25.3        63.3
std            43.5          21.5        33.8               42.1        60.8
min             0.0           0.0         0.0                0.0        15.0
25%             0.0           0.0         0.0                0.0        26.0
50%             2.0           0.0         2.0                8.0        43.0
75%            21.0           0.0        15.0               33.0        79.0
max         2,436.0       1,352.0     1,357.0            1,316.0     2,461.0

#### Encotrar datos faltantes por columnas

In [17]:
delayFlights.isnull().sum()

Month                     0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

#### Crear nuevas columnas
Una columna marca la velocidad media de cada vuelo dividiendo la distancia de vuelo en la velocidad y multiplando por 60 para pasar de minutos a horas
La otro columna avisa si los vuelos tuvieron retraso o no y clasifica los mismos en cortos, normales y largos.

In [18]:
#Creo una columan de velocidad mmedia de vuelo millas/h
             
delayFlights["MeanFlight/h"] = delayFlights['Distance']/delayFlights['AirTime']*60


In [19]:
delayFlights[['Distance', 'AirTime', 'MeanFlight/h']].head()

Distance  AirTime  MeanFlight/h
0       810    116.0         419.0
1       810    113.0         430.1
2       515     76.0         406.6
4       515     77.0         401.3
5       688     87.0         474.5

In [20]:
#Creo columnas que marquen 

conditions = [
    (delayFlights['ArrDelay'] < 15),
    (delayFlights['ArrDelay'] > 15) & (delayFlights['ArrDelay'] < 60),
    (delayFlights['ArrDelay'] > 60) & (delayFlights['ArrDelay'] < 90),
    (delayFlights['ArrDelay'] > 90),
    ]

# create a list of the values we want to assign for each condition
values = ['onTime','shortDelay' ,'delay' , 'longDelay']

# create a new column and use np.select to assign values to it using our lists as arguments
delayFlights['TypesOfDeley'] = np.select(conditions, values, default='Not Specified')
delayFlights[['ArrDelay', 'TypesOfDeley']]

ArrDelay TypesOfDeley
0           -14.0       onTime
1             2.0       onTime
2            14.0       onTime
4            34.0   shortDelay
5            11.0       onTime
...           ...          ...
7009710      25.0   shortDelay
7009717      75.0        delay
7009718      99.0    longDelay
7009726       9.0       onTime
7009727      -5.0       onTime

[1936758 rows x 2 columns]

#### Ver aerolineas con mayores retrasos

Todas las companías con sus minutos de demoras

In [21]:
#agrupo las companías en companies
companies = delayFlights.groupby('UniqueCarrier')

#imprimo todas las companías con sus acumulados de minutos de retrazo
companies['ArrDelay'].agg(np.sum)



UniqueCarrier
9E    2,420,468.0
AA    8,889,066.0
AQ       15,814.0
AS    1,406,735.0
B6    3,025,749.0
CO    4,045,932.0
DL    4,535,644.0
EV    3,888,131.0
F9      788,549.0
FL    3,100,150.0
HA      255,613.0
MQ    6,396,704.0
NW    3,462,075.0
OH    2,675,993.0
OO    5,978,936.0
UA    6,733,013.0
US    3,571,867.0
WN   11,319,092.0
XE    5,176,042.0
YV    3,691,461.0
Name: ArrDelay, dtype: float64

Todas las companías con sus minutos de demoras ORDENADAS de mayor a menor

In [22]:
delayCompanies = companies['ArrDelay'].agg(np.sum)

delayCompanies.sort_values(ascending = False)

UniqueCarrier
WN   11,319,092.0
AA    8,889,066.0
UA    6,733,013.0
MQ    6,396,704.0
OO    5,978,936.0
XE    5,176,042.0
DL    4,535,644.0
CO    4,045,932.0
EV    3,888,131.0
YV    3,691,461.0
US    3,571,867.0
NW    3,462,075.0
FL    3,100,150.0
B6    3,025,749.0
OH    2,675,993.0
9E    2,420,468.0
AS    1,406,735.0
F9      788,549.0
HA      255,613.0
AQ       15,814.0
Name: ArrDelay, dtype: float64

Las companías ordenadas de mayor a menor según el promedio de demoras de sus vuelos

In [23]:
delayCompanies = companies['ArrDelay'].agg(np.mean)

delayCompanies.sort_values(ascending = False)

UniqueCarrier
YV   55.3
B6   55.1
OH   51.0
XE   50.2
UA   47.8
EV   47.6
9E   46.9
AA   46.6
OO   45.4
MQ   45.3
NW   43.9
FL   43.7
CO   40.6
DL   39.9
US   36.4
AS   36.1
HA   34.2
WN   30.1
F9   27.9
AQ   21.3
Name: ArrDelay, dtype: float64

#### Vuelos más largo y más cortos 

In [24]:
#creo un DF sacando las distancias duplicadas
tempDeleyFlights = delayFlights.drop_duplicates(subset = ['Distance'])

#hago que me ordene las distancias de mayor a menor
distanceFlights = tempDeleyFlights.sort_values('Distance', ascending =False)

#Imprimo las diez distancias más largas
distanceFlights[['Dest', 'Origin', 'Distance']].head(10)


Dest Origin  Distance
564683   HNL    EWR      4962
305099   ATL    HNL      4502
218178   HNL    ORD      4243
218265   ORD    KOA      4213
218291   OGG    ORD      4184
438603   MSP    HNL      3972
563458   HNL    IAH      3904
477240   HNL    DFW      3784
477278   DFW    OGG      3711
2693079  ATL    ANC      3417

In [25]:
##Imprimo las diez distancias más cortas

distanceFlights[['Dest', 'Origin', 'Distance']].tail(10)


Dest Origin  Distance
3162610  BOS    MHT        45
1960669  GRR    MKG        41
3159197  TPA    SRQ        40
3701794  COS    PUB        37
169813   SNA    LAX        36
2484976  ONT    LGB        35
533827   WRG    PSG        31
550589   HPN    JFK        30
151660   DCA    IAD        24
2547298  LGA    JFK        11

### Ejercicio 3
Exporta el data set net i amb les noves columnes a Excel.

In [40]:
#guardo el csv para usar en el sprint 3
delayFlights.to_csv('DelayedFlightsLIMPIO.csv')



In [57]:
#Hago más pequeño el DF para que me quepa en un excel

delayFlights1 = delayFlights.loc[0:1000000].copy()

delayFlights1

delayFlights1.to_excel('DelayFligthsSHORT.xlsx', index=False)